polyfills/device-storage/device_storage.js

```diff
@@ -0,0 +1,260 @@
+// DeviceStorage polyfill!
+// https://developer.mozilla.org/en-US/docs/Web/API/Device_Storage_API
+
+// I have to define the service protocol, and this is as good place as any. The
+// protocol will be as follows:
+// * Each request will have a numeric request id, set by the client
+// * The answers will include the request id, so the client can know to which
+//   request the answer corresponds
+// * The request id will also act as identifier for remote objects
+// (deviceStorage)
+
+// For all the operations over a DeviceStorage object:
+//  * Request: { id: requestId,
+//               data: {
+//                 operation: methodName,
+//                 params: params,
+//                 deviceStorageId: deviceStorageId,
+//                                  (only non getStorageDevice operation)
+//                 storageName: storageName (only getStorageDevice operation)
+//               }
+//             }
+// * Answer: { id: requestId,
+//             result|error: Whatever }
+
+// For the onchange operation
+//  Request: { id: requestId,
+//             data: {
+//             operation: 'onchange'
+//           },
+// Answer: Will be invoked when there's activity with
+// the DeviceStorageChangeEvent (like) object:
+//    { id: requestId,
+//      data: DeviceStorageChangeEvent }
+
+
+(function(window) {
+
+  function debug(text) {
+    console.log('*-*-*- DeviceStorage PF: ' + text);
+  }
+
+  // TODO think about how to use the real api when available
+  /*if (window.navigator.mozSettings) {
+    // Hmm it's already available... so let's just use it and be done with it
+    return;
+  }*/
+
+  var realGetDeviceStorage = navigator.getDeviceStorage.bind(navigator);
+  var realGetDeviceStorages = navigator.getDeviceStorages.bind(navigator);
+
+  // Wishful thinking at the moment...
+  const DEVICESTORAGE_SERVICE = 'https://devicestorageservice.gaiamobile.org';
+
+  // It's nice being monothread...
+  var _currentRequestId = 1;
+  var _deviceStorageAttributes = ['canBeFormatted', 'canBeMounted',
+    'canBeShared', 'default', 'isRemovable', 'storageName'];
+
+  // DeviceStorage polyfill..
+  function FakeDeviceStorage(type, realDeviceStorage) {
+
+    var _resolve, _reject;
+    var _deviceStorage = new Promise((resolve, reject) => {
+      _resolve = resolve;
+      _reject = reject;
+    });
+
+    // This will hold the remote object,
+    // once the DeviceStorage is actually created..
+    var _deviceStorageId = null;
+    _deviceStorage.then(id => _deviceStorageId = id);
+
+    function _createAndQueueRequest(data) {
+      var request = new FakeDOMRequest(++_currentRequestId, data);
+      _sendRequest(request);
+      return request;
+    }
+
+    function _createAndQueueDOMCursorRequest(data) {
+      var request = new FakeDOMCursorRequest(++_currentRequestId, data);
+      _sendRequest(request);
+      return request;
+    }
+
```

Comment:
You can join both of those functions into one if you pass the type of object (the constructor in other words) as a parameter. You can even define a function template and two specific binds if you don't want to pass the parameter explicitly. The less duplicated code, the better :)